AX: lineTextMarkerRangeForTextMarker returns too much text for some blank lines

minorninth opened the pull request:
#### ca993564405e47a71e87c0ef6f2605fb0ec88911
<pre>
AX: lineTextMarkerRangeForTextMarker returns too much text for some blank lines
<a href="https://bugs.webkit.org/show_bug.cgi?id=277489">https://bugs.webkit.org/show_bug.cgi?id=277489</a>
<a href="https://rdar.apple.com/132989539">rdar://132989539</a>

Reviewed by Chris Fleizach.

In <a href="https://bugs.webkit.org/show_bug.cgi?id=275870">https://bugs.webkit.org/show_bug.cgi?id=275870</a> (AX:
lineTextMarkerRangeForTextMarker fails on first character of line in
Google Docs) we fixed an issue where lineTextMarkerRangeForTextMarker
failed to return a range at all.

This fix resulted in a regression where now it returns a range that&apos;s
too large, for some blank lines. One symptom was that when using
VoiceOver to arrow through editable text, landing on a blank line
would sometimes read the subsequent line rather than just &quot;newline&quot;.

* LayoutTests/accessibility/mac/line-text-marker-range-for-text-marker-expected.txt:
* LayoutTests/accessibility/mac/line-text-marker-range-for-text-marker.html:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::lineRangeForPosition const):

Canonical link: <a href="https://commits.webkit.org/281724@main">https://commits.webkit.org/281724@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/395e6c11fb46ef590e5779c86c4c8b5df011bd32

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60792 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40151 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13368 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64723 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11339 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62922 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47827 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11614 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/49163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7873 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62826 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37391 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52670 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/29994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34078 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9896 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10252 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55916 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10194 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66452 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4736 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10023 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56526 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4757 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52638 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56707 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13559 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3932 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/35956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37038 "Failed to checkout and rebase branch from PR 31601") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38131 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -1612,7 +1613,18 @@ std::optional<SimpleRange> AccessibilityObject::rangeForCharacterRange(const Cha
+    if (endPosition.isNull() || endPosition < startPosition) {
+        // When endOfLine fails to return a plausible result, try nextLineEndPosition, which is more robust, but ensure it doesn't return a result from a subsequent line.
+        VisiblePosition nextLineEnd = nextLineEndPosition(startPosition);
+        while (!nextLineEnd.isNull() && nextLineEnd > startPosition && lineForPosition(nextLineEnd) != lineForPosition(startPosition))
```
I think `lineForPosition` iterates every line until the nearest editable root, which might be really expensive (unless I'm reading that function wrong). And we do that twice here, in a loop. Would `inSameLine(nextLineEnd, startPosition)` work? Though one caveat is that `inSameLine` confusingly returns false if the positions are rendered on the same line, but one is editable (wrapped in `<span contenteditable="true">`) while the other isn't (e.g. wrapped by `<span contenteditable="false">`), and Google docs does have markup that can exhibit this (see the PR description of https://github.com/WebKit/WebKit/pull/30862).

minorninth replied:
Great call, let me see if isSameLine works or not.

minorninth replied:
inSameLine works great, done.

minorninth committed:
AX: lineTextMarkerRangeForTextMarker returns too much text for some blank lines
https://bugs.webkit.org/show_bug.cgi?id=277489
rdar://132989539

Reviewed by Chris Fleizach.

In https://bugs.webkit.org/show_bug.cgi?id=275870 (AX:
lineTextMarkerRangeForTextMarker fails on first character of line in
Google Docs) we fixed an issue where lineTextMarkerRangeForTextMarker
failed to return a range at all.

This fix resulted in a regression where now it returns a range that's
too large, for some blank lines. One symptom was that when using
VoiceOver to arrow through editable text, landing on a blank line
would sometimes read the subsequent line rather than just "newline".

* LayoutTests/accessibility/mac/line-text-marker-range-for-text-marker-expected.txt:
* LayoutTests/accessibility/mac/line-text-marker-range-for-text-marker.html:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::lineRangeForPosition const):

Canonical link: https://commits.webkit.org/281724@main
